### PR TITLE
ls: implement --group-directories-first

### LINKED ISSUE #3519: Implement `ls --group-directories-first`
Tested by gnu/tests/ls/group-dirs.sh  
```
       --group-directories-first
              group directories before files;

              can   be  augmented  with  a  --sort  option,  but  any  use  of
              --sort=none (-U) disables grouping
```

### PR DESCRIPTION
**Pull request by thomasqueirozb**
Closes #3519

### CODE REVIEW

**Comment by sylvestre** [comment]
Cool stuff. Could you please add tests? thanks

**Comment by thomasqueirozb** [comment]
> Cool stuff. Could you please add tests? thanks

I did it but I went down a rabbit hole. I found out that `ls -a` doesn't work the same as the GNU version. The `.` and `..` directories are always first, no matter what. If sorts are applied, `-r` is specified or `--group-directories-first` is used, the location of `.` and `..` shouldn't always be on the top.

Now with the changes I made it mostly behaves the same (I think). But this created one unfortunate side effect: `ls -a1v` now places `..` at the end of the list and `.` at the begging. The GNU version places both `.` and `..` at the begging. I couldn't find an easy fix for this so I just changed the test.

File: tests/by-util/test_ls.rs [hunk 1]
```diff
@@ -1301,6 +1301,59 @@ fn test_ls_deref() {
...
 
+#[test]
+fn test_ls_group_directories_first() {
+    let scene = TestScenario::new(util_name!());
+    let at = &scene.fixtures;
+    let mut filenames = ["file1", "file2", "anotherFile", "abc", "xxx", "zzz"];
+    for filename in filenames {
+        at.touch(filename);
+    }
+    filenames.sort();
```
Comment (sylvestre) [inline tests/by-util/test_ls.rs:1312]: please fix the clippy warning

Comment (thomasqueirozb) [reply tests/by-util/test_ls.rs]: Just did

**Comment by sylvestre** [comment]
>  I went down a rabbit hole. 

Welcome to the Rust/coreutils ;)

**Comment by thomasqueirozb** [comment]
An unrelated test failed. Maybe re-run that step of the CI?

**Comment by sylvestre** [comment]
bravo 
`Warning: Congrats! The gnu test tests/ls/group-dirs is no longer failing!`

**Pull request merged**